composites/Plugin/SnippetPreview/components/SnippetPreview.js

```diff
@@ -1,17 +1,597 @@
  *
- * @returns {ReactElement} The SnippetPreview component.
+ * @param {string} locale The locale.
+ * @param {string} keyword The keyword.
+ * @param {string} text The text in which to highlight a keyword.
+ *
+ * @returns {ReactElement} React elements to be rendered.
  */
-export default function SnippetPreview() {
-	return <SnippetPreviewDiv/>;
+function highlightKeyword( locale, keyword, text ) {
+	// Match keyword case-insensitively.
+	const keywordMatcher = createWordRegex( keyword, "", false );
+
+	text = text.replace( keywordMatcher, function( keyword ) {
+		return `{{strong}}${ keyword }{{/strong}}`;
+	} );
+
+	// Transliterate the keyword for highlighting
+	const transliteratedKeyword = transliterate( keyword, locale );
+	if ( transliteratedKeyword !== keyword ) {
+		const transliteratedKeywordMatcher = createWordRegex( transliteratedKeyword, "", false );
+		text = text.replace( transliteratedKeywordMatcher, function( keyword ) {
+			return `{{strong}}${ keyword }{{/strong}}`;
+		} );
+	}
+
+	return interpolateComponents( {
+		mixedString: text,
+		components: { strong: <strong /> },
+	} );
+}
+
+export default class SnippetPreview extends Component {
+	/**
+	 * Renders the SnippetPreview component.
+	 *
+	 * @param {Object} props The passed props.
+	 * @param {string} props.title                  The title tag.
+	 * @param {string} props.url                    The URL of the page for which to generate a snippet.
+	 * @param {string} props.description            The meta description.
+	 * @param {string} props.keyword                The keyword for the page.
+	 * @param {string} props.isDescriptionGenerated Whether the description was generated.
+	 * @param {string} props.locale                 The locale of the page.
+	 *
+	 * @returns {ReactElement} The SnippetPreview component.
+	 */
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			title: props.title,
+			description: props.description,
+		};
+
+		this.setTitleRef       = this.setTitleRef.bind( this );
+		this.setDescriptionRef = this.setDescriptionRef.bind( this );
+	}
+
+	setTitleRef( titleElement ) {
+		this._titleElement = titleElement;
+	}
+
+	setDescriptionRef( descriptionElement ) {
+		this._descriptionElement = descriptionElement;
+	}
+
+	/**
+	 * Returns whether an element has content that doesn't fit.
+	 *
+	 * Has a leeway of 2 to make sure weird measurements don't cause an infinite
+	 * loop.
+	 *
+	 * @param {HTMLElement} element The element to check.
+	 *
+	 * @returns {boolean} Whether it has content that doesn't fit.
+	 */
+	hasOverflowedContent( element ) {
+		return Math.abs( element.clientHeight - element.scrollHeight ) >= 2;
+	}
+
+	/**
+	 * Set the title in the state so it fits in two lines.
+	 *
+	 * @returns {void}
+	 */
+	fitTitle() {
+		const titleElement = this._titleElement;
+
+		// When the title is 600 pixels it approximately fits 200 characters.
+		const TWO_LINES_OF_CHARACTERS_PER_WIDTH = 3;
```

Comment:
This variable name is not very clear. Shouldn't this be something like pixels_per_character?

Reply:
Makes sense, I've also expanded the explanation.